tests/rust-plugins/swc_internal_plugin/src/lib.rs

```diff
@@ -23,35 +24,18 @@ impl VisitMut for ConsoleOutputReplacer {
     }
 }
 
-#[no_mangle]
-// TODO:
-// - this is not complete signature, need to take config_json
-// - swc_plugin macro to support ergonomic interfaces
-// - better developer experience: println / dbg!() doesn't emit anything
-pub fn process(ast_ptr: *mut u8, len: i32) -> (i32, i32) {
-    // Read raw serialized bytes from wasm's memory space. Host (SWC) should
-    // allocate memory, copy bytes and pass ptr to plugin.
-    let raw_serialized_bytes =
-        unsafe { std::slice::from_raw_parts(ast_ptr, len.try_into().unwrap()) };
-
-    // Reconstruct SerializedProgram from raw bytes
-    let serialized_program = Serialized::new_for_plugin(raw_serialized_bytes, len);
-
-    // TODO: Returning error pointer instead of unwrap
-    let program: Program =
-        Serialized::deserialize(&serialized_program).expect("Should able to deserialize");
-
-    // Actual plugin transformation
+/// An example plugin function with macro support.
+/// `plugin_module` macro interop pointers into deserialized structs, as well as
+/// returning ptr back to host.
+///
+/// It is possible to opt out from macro by writing transform fn manually via
+/// `__plugin_process_impl(*const u8, i32, *const u8, i32) -> (i32,i32)`
+/// if plugin need to handle low-level ptr directly.
+///
+/// TODO:
+/// - better developer experience: println / dbg!() doesn't emit anything
+#[plugin_module]
+pub fn process(program: Program, _plugin_config: String) -> Result<Program, std::fmt::Error> {
```

Comment:
Now we're back to original signature passes `Program` and `config` for both and macro hides all pointer interop to reconstruct `Program`.

- config is `string`: I was thinking to have some strict typed config, then realized this is config for the each plugin: plugin should able to change interface to config without changing swc's core. In result, this is a string and plugin author need to decompose as needed.

Few open questions

- should there be some deserialization helper for config into json (or other data types) or we should leave it to userland? (TBD for now)
- will we need any other metadata to be passed? or entire swc config? 
- Return value type: especially error handling is not defined yet, so `std:fmt::Error` in here is stub. Need to dig bit more to serialize error to pass into host.

Reply:
Actually, if we want to support babel-like configuration for plugins, the config can be an arbitrary type.

Type used in `extern "C"` should be fixed, but we can sielently serialize/deserialize for plugin authors. (Maybe by sing serde_json)

I'm not sure if this is good enough.

Reply:
Instead of passing information via `extern "C"`, we can allow plugin authors request for the required information using `imports! {}` macro I guess?
Of course it will require some work, but actually there are really lots of metadata and some of them are quite hard to pass.

Reply:
Since we get json-string, we can convert / deserialize as we want. But I'm not sure about ergonomics at this moment. Rather to try get early conclusion let's start from string first? in worst cases we could interop into something similar to serde_json:value and let user dynamically look up values.

Reply:
> information using imports! {} macro I guess

imports! macro is to expose host fn into guest, isn't it?

Reply:
I think starting with `String` is a good idea.

Reply:
not sure if tls need to be involved here since this lives in wasm's memory space, but as you said it may worth consider acquiring necessary type information to be reconstructed into plugin fn. 

I'd like to suggest this as open q for now, with getting some feedback once we release experimental release for the plugin.

Reply:
I agree. The only thing I worry about is `swc_common::errorrs::Handler` but it will be quite easy to implement support for it.